### PR TITLE
Support templating Openstack clusters as App CRs

### PR DESCRIPTION
**Pull request by ljakimczuk**
### Description

The `kubectl gs template cluster --provider openstack` is now able to output clusters as App CRs. It does that by running two additional commands underneath, one for templating the [cluste-openstack](https://github.com/giantswarm/cluster-openstack) app and another for templating the [default-apps-openstack](https://github.com/giantswarm/default-apps-openstack) app.

In order to accomplished that, the below flags have been introduced:
* `--cluster-topology`, set to `false` by default, when used cluster will be templated as App CR,
* `--cluster-user-configmap`, the current command supports multiple flags for customizing cluster templating, though in the case of App CRs, the configuration must be provided in a form of `values.yaml`. We could try to translate the existing cluster flags to `values.yaml` before running the `template app` commands, and in fact I tried that at the beginning, though it quickly started to look messy and felt counterintuitive. Working with the `values.yaml` directly seems much more easier and flexible in configuration to the users, so I introduced this extra flag,
* `--default-apps-user-configmap`, the same reasoning as above applies here,
* `--cluster-app-version`, allows cluster app version selection,
* `--default-apps-app-version`, allows default apps app version selection.

### Examples

1. `$ kubectl gs template cluster --provider openstack --cluster-topology --organization giantswarm-2`

Output:
```yaml
---
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 0.0.0
  name: ox0f8-cluster
  namespace: org-giantswarm-2
spec:
  catalog: control-plane-catalog
  config:
    configMap:
      name: ""
      namespace: ""
    secret:
      name: ""
      namespace: ""
  kubeConfig:
    context:
      name: ""
    inCluster: true
    secret:
      name: ""
      namespace: ""
  name: cluster-openstack
  namespace: org-giantswarm-2
  version: 0.1.0
---
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 0.0.0
  name: ox0f8-default-apps
  namespace: org-giantswarm-2
spec:
  catalog: default
  config:
    configMap:
      name: ""
      namespace: ""
    secret:
      name: ""
      namespace: ""
  kubeConfig:
    context:
      name: ""
    inCluster: true
    secret:
      name: ""
      namespace: ""
  name: default-apps-openstack
  namespace: org-giantswarm-2
  version: 0.1.0
```

2. `$ kubectl gs template cluster --provider openstack --cluster-topology --organization giantswarm-2 --cluster-user-configmap /tmp/cm.yaml`

```yaml
---
apiVersion: v1
data:
  values: |
    clusterDescription: "My test cluster"
    organization: "giantswarm-2"
    cloudConfig: cloud-config
    cloudName: openstack

    kubernetesVersion: 1.20.9
    releaseVersion: 20.0.0-alpha1

    nodeClasses:
    - name: small
      machineFlavor: n1.small
      diskSize: 20

    controlPlane:
      class: default
      replicas: 3

    nodePools:
    - name: default
      class: default
      replicas: 3
kind: ConfigMap
metadata:
  creationTimestamp: null
  name: zi95g-cluster-userconfig
  namespace: org-giantswarm-2
---
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 0.0.0
  name: zi95g-cluster
  namespace: org-giantswarm-2
spec:
  catalog: control-plane-catalog
  config:
    configMap:
      name: ""
      namespace: ""
    secret:
      name: ""
      namespace: ""
  kubeConfig:
    context:
      name: ""
    inCluster: true
    secret:
      name: ""
      namespace: ""
  name: cluster-openstack
  namespace: org-giantswarm-2
  userConfig:
    configMap:
      name: zi95g-cluster-userconfig
      namespace: org-giantswarm-2
  version: 0.1.0
---
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 0.0.0
  name: zi95g-default-apps
  namespace: org-giantswarm-2
spec:
  catalog: default
  config:
    configMap:
      name: ""
      namespace: ""
    secret:
      name: ""
      namespace: ""
  kubeConfig:
    context:
      name: ""
    inCluster: true
    secret:
      name: ""
      namespace: ""
  name: default-apps-openstack
  namespace: org-giantswarm-2
  version: 0.1.0
```

3. `$ kubectl gs template cluster --provider openstack --organization giantswarm-2 --release 20.0.0-alpha1 --node-machine-flavor=n1.small` (some output has been cut for brevity)

```yaml
---
apiVersion: cluster.x-k8s.io/v1alpha4
kind: Cluster
metadata:
  annotations:
    cluster.giantswarm.io/description: ""
  labels:
    cluster.x-k8s.io/cluster-name: kz4j9
    giantswarm.io/cluster: kz4j9
    giantswarm.io/organization: giantswarm-2
    release.giantswarm.io/version: 20.0.0-alpha1
  name: kz4j9
  namespace: org-giantswarm-2
spec:
  controlPlaneEndpoint:
    host: ""
    port: 0
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
    kind: KubeadmControlPlane
    name: kz4j9
    namespace: org-giantswarm-2
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
    kind: OpenStackCluster
    name: kz4j9
    namespace: org-giantswarm-2
status:
  infrastructureReady: false
---
...
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: OpenStackMachineTemplate
metadata:
  labels:
    cluster.x-k8s.io/cluster-name: kz4j9
    giantswarm.io/cluster: kz4j9
    giantswarm.io/organization: giantswarm-2
    release.giantswarm.io/version: 20.0.0-alpha1
  name: kz4j9
  namespace: org-giantswarm-2
spec:
  template:
    spec:
      cloudName: openstack
      flavor: n1.small
      identityRef:
        kind: Secret
        name: cloud-config
      image: ubuntu-2004-kube-v1.20.9
      rootVolume: {}
```

### CODE REVIEW

File: cmd/template/cluster/provider/capo.go
Comment (MarcelMue) [inline cmd/template/cluster/provider/capo.go:79]: This should most likely be configurable.

Comment (ljakimczuk) [reply cmd/template/cluster/provider/capo.go:79]: The`version` is hard coded just for a start and I will add a flag to configure it. And did you mean the other fields as well?

Comment (ljakimczuk) [reply cmd/template/cluster/provider/capo.go:79]: Sure, got it. I must also add another `user-configmap` for the app of apps

File: cmd/template/app/runner.go
Comment (rossf7) [inline cmd/template/app/runner.go:21]: Rather than exporting the runner I'd prefer more of the logic is moved to `pkg/template/app`.
Can you take a look and see how hard that would be?

Reason being that multiple teams need to work on `kubectl gs` so it's good to keep the code as consistent as possible. 

Also me may need to support more types of apps in future. e.g. an app of apps for the security stack.

File: cmd/template/cluster/flag.go
Comment (rossf7) [inline cmd/template/cluster/flag.go:18]:
```suggestion
	flagTemplateType  = "template-type"
```

Flags are not that long so I think its better to align the constants with the full name.

File: cmd/template/cluster/flag.go
Comment (rossf7) [inline cmd/template/cluster/flag.go:29]:
```suggestion
	flagClusterUserConfigMap      = "cluster-user-configmap"
	flagDefaultAppsUserConfigMap  = "default-apps-user-configmap"
```

File: cmd/template/cluster/flag.go
Comment (rossf7) [inline cmd/template/cluster/flag.go:112]:
```suggestion
	cmd.Flags().StringVar(&f.TplType, flagTplType, "raw", "Type of the template. Could be one of: raw, app. If application, cluster will be templated as an App CR. (default raw)")
```

We use apps in our docs already. Let's see what sig UX says later 😅 but I prefer just `app`.

File: cmd/template/cluster/flag.go
Comment (ljakimczuk) [reply cmd/template/cluster/flag.go]: Makes sense, done it

File: cmd/template/cluster/flag.go
Comment (ljakimczuk) [reply cmd/template/cluster/flag.go]: I agree `app` looks better 😄

File: cmd/template/app/runner.go
Comment (ljakimczuk) [reply cmd/template/app/runner.go]: @rossf7  Agree that exporting apps introduced some inconsistency, it may hurt aesthetics 😄 I tried to re-organize the code a bit, to move some parts to the `package/template/app` and thus avoid that much code repetition, though I failed to fully accomplish the goal. At the moment the `pkg/template/app` provides some structs and functions, we relay on the returned values to modify instances of these structs in the `cmd/template/app`, that's why a good dose of code lives in the `cmd/template/app` and the `cmd/template/cluster/provider/capo`. I believe if we want to move more of it into the `package/template/app` we will need to make classes out of these structs, and make the functions to be their methods, but I am not sure if that's not too big of a change.

File: cmd/template/cluster/flag.go
Comment (rossf7) [inline cmd/template/cluster/flag.go:112]:
```suggestion
	cmd.Flags().StringVar(&f.TemplateType, flagTemplateType, "raw", "Type of the template. Could be one of: raw, app. If app, cluster will be templated as an App CR. (default raw)")
```

File: cmd/template/cluster/provider/capo.go
Comment (rossf7) [inline cmd/template/cluster/provider/capo.go:92]:
```suggestion
		Catalog:   "default",
```

The app is only pushed to the default catalog right now. Since that's where the child apps are.

You can check this is in the Circle [config](https://github.com/giantswarm/default-apps-openstack/blob/3a994712a58b2183737201e85cbcbabee8498321/.circleci/config.yml#L12)

File: cmd/template/cluster/provider/common.go
Comment (rossf7) [inline cmd/template/cluster/provider/common.go:58]:
```suggestion
        // App settings.
	ClusterAppUserConfigMap     string
	ClusterAppVersion           string
	DefaultAppsAppUserConfigMap string
	DefaultAppsAppVersion       string
```

File: cmd/template/cluster/runner.go
Comment (rossf7) [inline cmd/template/cluster/runner.go:85]:
```suggestion
			ClusterAppVersion:           r.flag.ClusterApp.ClusterAppVersion,
			ClusterAppUserConfigMap:     r.flag.ClusterApp.ClusterUserConfigMap,
			DefaultAppsAppVersion:       r.flag.ClusterApp.DefaultAppsAppVersion,
			DefaultAppsAppUserConfigMap: r.flag.ClusterApp.DefaultAppsUserConfigMap,
```

File: cmd/template/cluster/provider/capo.go
Comment (rossf7) [inline cmd/template/cluster/provider/capo.go:82]:
```suggestion
		AppName:   fmt.Sprintf("%s-cluster", config.Name),
```

File: cmd/template/cluster/provider/capo.go
Comment (rossf7) [inline cmd/template/cluster/provider/capo.go:91]:
```suggestion
		AppName:   fmt.Sprintf("%s-default-apps", config.Name),
```

File: cmd/template/app/runner.go
Comment (rossf7) [reply cmd/template/app/runner.go]: Thx! The structure looks a lot better ❤️ 

The main thing was I didn't want the runner to be exported. I think this is fine for now we can always revisit when we add support for more providers.

**Review by rossf7 — COMMENTED** [review]
@ljakimczuk This is looking good but CI isn't happy. Can you check and appease the Circle gods? :)

File: cmd/template/app/runner.go
Comment (ljakimczuk) [reply cmd/template/app/runner.go]: Perfect, good to hear that, I made the suggested changes and fixed import so that CI's happy.

File: cmd/template/cluster/provider/capo.go
Comment (ljakimczuk) [reply cmd/template/cluster/provider/capo.go]: Ok, the fields now have their own flags

File: cmd/template/cluster/flag.go
Comment (rossf7) [inline cmd/template/cluster/flag.go:29]: Nit but could you order the flags in the group in alpha order?

That looks tidier to me 🤓

**Review by rossf7 — COMMENTED** [review]
Nice<3 this is looking good. Just some final nits on ordering.

We should get a review from Sig UX before merging but we can start with a review from the team.

We'll also need a docs PR 📚 https://docs.giantswarm.io/ui-api/kubectl-gs/template-cluster/

File: cmd/template/cluster/flag.go
Comment (ljakimczuk) [reply cmd/template/cluster/flag.go:29]: Sure thing!

**Comment by ljakimczuk** [comment]
> We'll also need a docs PR.

I remember about that, but let's maybe wait for that until we get more reviews, just in case we have more changes coming in. WDYT?

**Comment by rossf7** [comment]
> maybe wait for that until we get more reviews, just in case we have more changes coming in. WDYT?

@ljakimczuk Yes that's fine 👍

File: cmd/template/cluster/flag.go
Comment (tfussell) [inline cmd/template/cluster/flag.go:133]: I think we want to hide these flags for now, see https://github.com/giantswarm/kubectl-gs/pull/540#issuecomment-977888540 and https://github.com/giantswarm/roadmap/issues/605. WDYT @giantswarm/team-rocket?

**Review by tfussell — APPROVED** [review]

File: cmd/template/cluster/flag.go
Comment (anvddriesch) [reply cmd/template/cluster/flag.go:133]: I would suggest this, too.

**Review by anvddriesch — COMMENTED** [review]
Cool stuff!
Please list all the changes to the `CHANGELOG`.

**Review by rossf7 — APPROVED** [review]
LGTM when @anvddriesch and team 🌈 are happy

Agree it makes sense to hide the openstack flags for now.

File: cmd/template/cluster/flag.go
Comment (ljakimczuk) [reply cmd/template/cluster/flag.go:133]: Ok, it's done

**Comment by ljakimczuk** [comment]
@rossf7 @anvddriesch @tfussell done, flags are now hidden and the changlog entry is there, let me know if I should make it more detailed.

**Review by anvddriesch — APPROVED** [review]
LGTM

**Comment by ljakimczuk** [comment]
@kopiczko not necessarily, it seemed more consistent to me to have this type, but it would work just as good without it. We can get rid of it or change this flag to something not suggesting multiple types, for example, `--as-app` or similar

**Comment by kopiczko** [comment]
> can get rid of it or change this flag to something not suggesting multiple types, for example, `--as-app` or similar

I thought we could replace existing templates and ditch them altogether.

**Comment by ljakimczuk** [comment]
I believe the original idea was to support both templates types, at least for some time since some users may be interested in templating raw resources, but maybe I got it wrong. @rossf7 WDYT?

**Comment by kopiczko** [comment]
Asked @cornelius-keller

**Comment by cornelius-keller** [comment]
I think currently it does not hurt and gives the customer the opportunity to create the clusters with the old method still, which might be useful in case there are problems with the new ones and we need to compare them to see what is different.

In the long run I agree with @kopiczko and it should disappear or be renamed, especially because customers can expect that the raw method templates  the same cluster as the app method only that it is the raw template and not wrapped into an app.

Maybe it will be useful to provide an option to get a raw template that is the same as included in the app, but that is a different story.

LGTM

**Comment by kopiczko** [comment]
It would be nice to at least generate raw templates using `helm template` to make sure they are always on par.

**Comment by ljakimczuk** [comment]
> It would be nice to at least generate raw templates using helm template to make sure they are always on par.

This may be misleading for those relying on the "old" version for the migration time being and as far as I understood that's indeed the case. But since `app` is mostly about adding `ClusterTopology`, @tfussell suggested we could change the flag to `--cluster-topology [true|false]`, @tfussell would you like to add something to that?

File: cmd/template/cluster/provider/capo.go
Comment (rossf7) [inline cmd/template/cluster/provider/capo.go:83]: @ljakimczuk @tfussell I found a problem. We use `.Release.Name` in the cluster helm chart which turns out badly :(

```
ross3-cluster   Provisioned   4m37s   v1.20.9
```

Should we remove the `-cluster` suffix here or in the helm chart?

Comment (ljakimczuk) [reply cmd/template/cluster/provider/capo.go:83]: I guess we could remove it from here, though App name may then be vague to the user, cause in some cases it would be auto-generated

**Comment by marians** [comment]
Haven't read through the entire discussion.

One minor thing that stands out to me is the duplication of the term `template` in the flag `--template-type`. If we need that flag, could it simply be `--type` ?

**Comment by ljakimczuk** [comment]
> One minor thing that stands out to me is the duplication of the term template in the flag --template-type. If we need that flag, could it simply be --type

I duplicated term `template` to emphasize the fact it's about the template type. I was afraid purpose of the `--type` alone may be vague having all the providers and different resources generated. 

@tfussell has already proposed to change this flag to something else, namely `--cluster-topology [true|false]`, that more closely describe the purpose of having an app for cluster, what do you think @marians ?

When we have an agreement I will push the changes.

**Comment by marians** [comment]
With a boolean flag, the default (when the flag is not used) should always be `false`. So `--cluster-topology false` would be the default. I'd like to make sure that the default makes sense.

What sort of output will that (`--cluster-topology false`) generate then, and what will  `--cluster-topology` or `--cluster-topology true` generate instead?

**Comment by ljakimczuk** [comment]
@marians The `--cluster-topology false` in  this scenario would behave the same way the `--template-type raw` (or lack of it) behaves. The `cluster-topology true` would behave the same way the `--template-type app` behaves right now.

**Review by marians — APPROVED** [review]

**Pull request merged**
